Add `.add()` method

therealklanni opened the pull request:
Implements sync and async add operations on object lists. Each method returns a new object-list instance to keep them immutable.

The async method is a partial implementation, pending adapter implementation, but can be used as is (it just gets immediately resolved for now). Once adapters are implemented, the async method can be updated without impacting anything using the current async implementation.

therealklanni committed:
Move record data to external json

therealklanni committed:
Basic add method implementation

therealklanni committed:
Ignore npm-debug.log

therealklanni committed:
Return new instance

therealklanni committed:
Revert "Move record data to external json"

This reverts commit c99d9114a14888862570f2bfeb82dbe4e489c59b.

therealklanni committed:
Add records to `.add()` test

therealklanni committed:
Remove unneeded assertion

therealklanni committed:
Implement async `.add()` operation

therealklanni committed:
Define `objectList` at top

ericelliott commented on index.js:
```diff
@@ -42,21 +56,52 @@ var
+      where: function (keyWhitelist) {
+        return where.apply(null, [collection, keyWhitelist]);
+      },
+      add: function (records, cb) {
```
This one shouldn't take records or callback. This is the method version which already has access to the collection via closure, and the callback gets hooked up via `.subscribe`.

therealklanni replied:
`cb` is a param because then I don't need to add `arguments[1]` onto the array passed to `add.apply`, instead I just name it. The extra param will be ignored if it's not an async invocation. `cb` will be the node-ized deferred object from Q when `add` is invoked as async.

The `records` param is the new record(s) being added. Let me know if this is not how you want it implemented, I can change it.

therealklanni replied:
See my comments at the bottom for a different approach. If you don't like the new approach, I can revert it.

ericelliott commented on index.js:
```diff
@@ -34,6 +37,17 @@ var
+    var newCollection = collection.concat(records);
+
+    if (cb) {
+      cb(null, newCollection);
```
shouldn't this `newCollection` value also be wrapped with `objectList`?

therealklanni replied:
I may have overlooked that. I'll go over it again and update the code. Thanks :+1:

ericelliott commented on test/index.js:
```diff
@@ -266,3 +266,136 @@ skip('.where() multiple clause', function (assert) {
+    });
+
+  result.add(newRecord);
+  result.subscribe(
```
This should fail because `result` should not be altered by `result.add()`.

If this works, it means that `.add()` mutates the `result` collection, right?

Instead, `result.add()` should return a new object-list that can be subscribed to.

therealklanni committed:
Wrap `newCollection` with objectList when invoking cb

therealklanni committed:
Add method should return new objectList instance

New objectList instance will provide subscribe method

therealklanni commented:
@ericelliott please let me know if ff48dd5 is more what you were thinking.

ericelliott commented on index.js:
```diff
@@ -42,21 +76,42 @@ var
+        return where.apply(null, [collection, keyWhitelist]);
+      },
+      add: function (records) {
+        return add.apply(null, [collection, records]);
```
Yes. =)

ericelliott commented on index.js:
```diff
@@ -42,21 +76,42 @@ var
+        return add.apply(null, [collection, records]);
+      },
+      push: function () {
+        return api.add.apply(null, [].slice.call(arguments));
```
Is `[].slice.call()` really needed here?

Did you try `return api.add.apply(null, arguments);`?

therealklanni replied:
I didn't because I thought I remembered that `arguments` doesn't work with `.apply()` since it's not a true array, but I'm probably misremembering. I'll try it without the slice.

ericelliott replied:
I'm pretty sure it does, but lots of programs try to manipulate the arguments before or after they get passed using stuff that isn't available on `arguments`, so it usually gets sliced -- but if you don't need to do any manipulation, you can just pass it as-is.

therealklanni replied:
Must be force of habit :) I'll update that.

ericelliott commented on index.js:
```diff
@@ -42,21 +76,42 @@ var
+  if (options.async) {
+    collection = options.list;
+
+    api.add = addAsync.bind(null, collection);
```
Do you propose we do this one at a time for every function? Is there a more generic way we could handle this?

therealklanni replied:
I'm open to suggestions :)

ericelliott replied:
How about some dreamcode? Let's rewrite the wrapper:

``` js
      add: function (records) {
        var version = options.async? 'async' : 'sync';
        return fnVersions[version].add.apply(null, [collection, records]);
      }
```

therealklanni commented:
It was passing, but only because I forgot to wrap the `newCollection` passed to `cb` with `objectList`. When I fixed that I also fixed the test so it would pass. See 0a5c109

therealklanni committed:
No need to slice

therealklanni committed:
Use `fnVersions` approach

therealklanni committed:
Add travis.yml

therealklanni committed:
Remove unused param

ericelliott commented:
Please run `npm run validate` and fix any issues you see.

therealklanni commented:
I see none ;)

therealklanni committed:
Add travis badge

ericelliott commented:
Odd... I see...

```
Erics-MacBook-Pro:object-list eric$ npm run validate

> object-list@1.1.1 validate /Users/eric/Dropbox/dev/object-list
> npm run lint && npm test && npm outdated --depth 0


> object-list@1.1.1 lint /Users/eric/Dropbox/dev/object-list
> jshint .

index.js: line 34, col 51, 'cb' is defined but never used.

1 error
```

therealklanni commented:
You probably didn't pull the latest ;)
